File: a3_finance/a3_finance/report/housing_loan_contribution_summary/housing_loan_contribution_summary.py
# Copyright (c) 2025, Acube and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import get_first_day, get_last_day


def execute(filters=None):
    columns = get_columns()
    data = get_data(filters)
    return columns, data


def get_columns():
    return [
        {"label": "Sl No", "fieldname": "sl_no", "fieldtype": "Int", "width": 50},
        {"label": "Employee Number", "fieldname": "employee_number", "fieldtype": "Data", "width": 120},
        {"label": "Employee Name", "fieldname": "employee_name", "fieldtype": "Data", "width": 200},
        {"label": "Phone No", "fieldname": "phone_no", "fieldtype": "Data", "width": 120},
        {"label": "Amount", "fieldname": "amount", "fieldtype": "Currency", "width": 120},
        {"label": "INSTITUTION/BANK", "fieldname": "institution", "fieldtype": "Data", "width": 180},
    ]


def get_data(filters):
    month = int(filters.get("month"))
    year = int(filters.get("year"))
    start_date = get_first_day(f"{year}-{month}-01")
    end_date = get_last_day(f"{year}-{month}-01")

    add_sal = frappe.get_all(
        "Additional Salary",
        filters={
            "salary_component": "Housing Loan",
            "is_recurring": 1,
            "disabled": 0,
            "docstatus": 1,
        },
        fields=["*"],
    )

    data = []
    sl_no = 1
    total_amount = 0

    for d in add_sal:
        # find salary slip for this employee in the given month
        salary_slip = frappe.db.sql(
            """
            SELECT name FROM `tabSalary Slip`
            WHERE employee = %s AND start_date >= %s AND end_date <= %s AND docstatus = 1
            LIMIT 1
            """,
            (d.employee, start_date, end_date),
            as_dict=True,
        )

        if salary_slip:
            loan_amount = frappe.db.sql(
                """
                SELECT amount FROM `tabSalary Detail`
                WHERE parent = %s AND salary_component = %s
                """,
                (salary_slip[0].name, "Housing Loan"),
                as_dict=True,
            )

            deducted_amount = loan_amount[0].amount if loan_amount else 0

            emp = frappe.db.get_value(
                "Employee",
                d.employee,
                ["employee_number", "employee_name", "cell_number"],
                as_dict=True,
            )

            # safely fetch bank name (skip if field not exists)
            bank_name = None
            try:
                bank_name = frappe.db.get_value("Employee", d.employee, "custom_bank_name")
            except Exception:
                pass

            data.append({
                "sl_no": sl_no,
                "employee_number": emp.employee_number if emp else "",
                "employee_name": emp.employee_name if emp else "",
                "phone_no": emp.cell_number if emp else "",
                "amount": deducted_amount,
                "institution": bank_name or "",
            })

            sl_no += 1
            total_amount += deducted_amount

    # add total row
    data.append({
        "sl_no": None,
        "employee_number": "",
        "employee_name": "Total",
        "phone_no": "",
        "amount": total_amount,
        "institution": "",
    })

    return data

File: a3_finance/a3_finance/report/housing_loan_contribution_summary/housing_loan_contribution_summary.js
// Copyright (c) 2025, Acube and contributors
// For license information, please see license.txt

frappe.query_reports["Housing Loan Contribution Summary"] = {

	onload: function (report) {

		// ---- PRINT BUTTON (ECR STYLE) ----
		report.page.add_inner_button("Print Report", function () {
			print_housing_loan_report(report);
		});
	},

	filters: [
		{
			fieldname: "month",
			label: "Month",
			fieldtype: "Select",
			options: ["01","02","03","04","05","06","07","08","09","10","11","12"],
			default: frappe.datetime.get_today().split("-")[1]
		},
		{
			fieldname: "year",
			label: "Year",
			fieldtype: "Int",
			default: frappe.datetime.get_today().split("-")[0]
		}
	],
printable_html: async function (report) {

	let filters = frappe.query_report.get_filter_values();
	let data = report.data || [];
	let columns = report.columns || [];

	const month_names = [
		"January","February","March","April","May","June",
		"July","August","September","October","November","December"
	];

	let month_name = month_names[parseInt(filters.month) - 1];

	let html = `
<style>
@media print {
	table { border-collapse: collapse; width: 100%; font-size: 11px; }
	th, td { border: 1px solid #000; padding: 4px; }
	th { text-align: center; font-weight: bold; }
	td { text-align: left; }
	td.num { text-align: right; }
	tr.total-row { font-weight: bold; }
}
</style>

<div style="text-align:center; margin-bottom:15px;">
	<div style="font-size:18px; font-weight:bold;">
		Housing Loan Contribution Summary
	</div>
	<div style="font-size:14px; margin-top:4px;">
		Month: <b>${month_name}</b> &nbsp; | &nbsp;
		Year: <b>${filters.year}</b>
	</div>
</div>

<table>
	<thead>
		<tr>
`;

	columns.forEach(col => {
		html += `<th>${col.label}</th>`;
	});

	html += `</tr></thead><tbody>`;

	data.forEach(row => {

		let is_total = row.employee_name === "Total" || row.employee === "Total";

		html += `<tr class="${is_total ? 'total-row' : ''}">`;

		columns.forEach(col => {
			let cls = col.fieldtype === "Currency" ? "num" : "";

			if (is_total && col.fieldname === "sl_no") {
				// Empty cell for total row
				html += `<td class="${cls}"></td>`;
			} else {
				let value = row[col.fieldname] ?? "";
				html += `<td class="${cls}">${frappe.format(value, col)}</td>`;
			}
		});

		html += `</tr>`;
	});

	html += `
	</tbody>
</table>

<p style="text-align:right; margin-top:25px; font-size:9px;">
	Printed on ${frappe.datetime.str_to_user(
		frappe.datetime.get_datetime_as_string()
	)}
</p>
`;

	return html;
}


};


// ---------------- CUSTOM PRINT FUNCTION ----------------
async function print_housing_loan_report(report) {

	let html = await frappe.query_reports["Housing Loan Contribution Summary"]
		.printable_html(report);

	let print_window = window.open("", "PRINT", "height=700,width=900");

	print_window.document.write(html);
	print_window.document.close();
	print_window.focus();
	print_window.print();
	print_window.close();
}
